File: host/websocket/prelude.js
// WebSocket web-API surface for gojs. This runs once at Install time and
// defines globalThis.WebSocket, delegating all I/O to the hidden native bridge
// globalThis.__gojs_ws (connect/send/close).
//
// Divergences from browsers: there is no Blob, so binary messages always arrive
// as an ArrayBuffer and binaryType defaults to "arraybuffer" (browsers default
// to "blob"). Everything else follows the WHATWG WebSocket interface.
(function () {
  "use strict";

  var WS = globalThis.__gojs_ws;

  var CONNECTING = 0;
  var OPEN = 1;
  var CLOSING = 2;
  var CLOSED = 3;

  var EVENT_TYPES = ["open", "message", "error", "close"];

  // A minimal Event-like object. MessageEvent carries `data`; CloseEvent carries
  // `code`, `reason`, `wasClean`; the error event carries `message`.
  function makeEvent(type, props) {
    var ev = { type: type, target: null };
    if (props) {
      for (var k in props) {
        if (Object.prototype.hasOwnProperty.call(props, k)) ev[k] = props[k];
      }
    }
    return ev;
  }

  class WebSocket {
    constructor(url, protocols) {
      if (url === undefined) {
        throw new TypeError("Failed to construct 'WebSocket': 1 argument required");
      }
      this.url = String(url);
      this._readyState = CONNECTING;
      this.protocol = "";
      this.extensions = "";
      this.bufferedAmount = 0;
      this.binaryType = "arraybuffer";
      this.onopen = null;
      this.onmessage = null;
      this.onerror = null;
      this.onclose = null;
      this._listeners = { open: [], message: [], error: [], close: [] };

      var protoArg = "";
      if (protocols !== undefined && protocols !== null) {
        if (Array.isArray(protocols)) protoArg = protocols.join(", ");
        else protoArg = String(protocols);
      }

      var self = this;
      this._handle = WS.connect(this.url, protoArg, function (type, a, b, c) {
        self._dispatch(type, a, b, c);
      });
    }

    get readyState() {
      return this._readyState;
    }

    _dispatch(type, a, b, c) {
      switch (type) {
        case "open":
          this._readyState = OPEN;
          this.protocol = a || "";
          this.extensions = b || "";
          this._fire("open", makeEvent("open"));
          break;
        case "message":
          this._fire("message", makeEvent("message", { data: a }));
          break;
        case "error":
          this._fire("error", makeEvent("error", { message: a || "" }));
          break;
        case "close":
          this._readyState = CLOSED;
          this._fire(
            "close",
            makeEvent("close", { code: a, reason: b || "", wasClean: !!c })
          );
          break;
      }
    }

    _fire(type, ev) {
      ev.target = this;
      var handler = this["on" + type];
      if (typeof handler === "function") {
        try {
          handler.call(this, ev);
        } catch (e) {
          // Listener exceptions must not break the dispatch, mirroring the DOM.
        }
      }
      var list = this._listeners[type];
      if (list) {
        var copy = list.slice();
        for (var i = 0; i < copy.length; i++) {
          try {
            copy[i].call(this, ev);
          } catch (e) {
            // swallow, as above
          }
        }
      }
    }

    send(data) {
      if (this._readyState === CONNECTING) {
        var err = new Error("Failed to execute 'send' on 'WebSocket': still in CONNECTING state");
        err.name = "InvalidStateError";
        throw err;
      }
      if (this._readyState !== OPEN) {
        // CLOSING or CLOSED: the spec drops the data silently.
        return;
      }
      WS.send(this._handle, data);
    }

    close(code, reason) {
      if (code !== undefined) {
        code = code | 0;
        if (code !== 1000 && !(code >= 3000 && code <= 4999)) {
          var err = new Error("Failed to execute 'close' on 'WebSocket': invalid code " + code);
          err.name = "InvalidAccessError";
          throw err;
        }
      }
      if (reason !== undefined && reason !== null) {
        reason = String(reason);
        // A close reason is limited to 123 bytes of UTF-8 (spec).
        if (reason.length > 123) {
          var e2 = new Error("Failed to execute 'close' on 'WebSocket': reason too long");
          e2.name = "SyntaxError";
          throw e2;
        }
      }
      if (this._readyState === CLOSING || this._readyState === CLOSED) {
        return;
      }
      this._readyState = CLOSING;
      WS.close(this._handle, code, reason);
    }

    addEventListener(type, listener) {
      if (EVENT_TYPES.indexOf(type) < 0 || typeof listener !== "function") return;
      var list = this._listeners[type];
      if (list.indexOf(listener) < 0) list.push(listener);
    }

    removeEventListener(type, listener) {
      var list = this._listeners[type];
      if (!list) return;
      var idx = list.indexOf(listener);
      if (idx >= 0) list.splice(idx, 1);
    }

    dispatchEvent(event) {
      if (!event || typeof event.type !== "string") return true;
      this._fire(event.type, event);
      return true;
    }
  }

  // readyState constants as static and instance (prototype) properties.
  WebSocket.CONNECTING = CONNECTING;
  WebSocket.OPEN = OPEN;
  WebSocket.CLOSING = CLOSING;
  WebSocket.CLOSED = CLOSED;
  WebSocket.prototype.CONNECTING = CONNECTING;
  WebSocket.prototype.OPEN = OPEN;
  WebSocket.prototype.CLOSING = CLOSING;
  WebSocket.prototype.CLOSED = CLOSED;

  globalThis.WebSocket = WebSocket;
})();

File: host/sse/prelude.go
package sse

// prelude is the JavaScript that defines globalThis.EventSource. It is loaded by
// Install via vm.RunString and delegates networking to the hidden natives under
// globalThis.__gojs_sse (installed by Install). Keeping the web-facing API in
// script lets us model the DOM EventTarget semantics (handler properties plus
// addEventListener/removeEventListener/dispatchEvent) without a native
// constructor factory.
const prelude = `
(function () {
  "use strict";
  const native = globalThis.__gojs_sse;

  const CONNECTING = 0;
  const OPEN = 1;
  const CLOSED = 2;

  class EventSource {
    constructor(url, config) {
      if (url === undefined || url === null) {
        throw new TypeError("Failed to construct 'EventSource': url is required");
      }
      this.url = String(url);
      this.withCredentials = !!(config && config.withCredentials);
      this.readyState = CONNECTING;
      this.onopen = null;
      this.onmessage = null;
      this.onerror = null;
      this._listeners = Object.create(null);
      const self = this;
      this._handle = native.connect(
        this.url,
        { withCredentials: this.withCredentials },
        function (kind, payload) { self._onNative(kind, payload); }
      );
    }

    addEventListener(type, listener) {
      if (typeof listener !== "function") return;
      type = String(type);
      const list = this._listeners[type] || (this._listeners[type] = []);
      if (list.indexOf(listener) === -1) list.push(listener);
    }

    removeEventListener(type, listener) {
      type = String(type);
      const list = this._listeners[type];
      if (!list) return;
      const i = list.indexOf(listener);
      if (i !== -1) list.splice(i, 1);
    }

    dispatchEvent(event) {
      const type = event && event.type;
      const handler = this["on" + type];
      if (typeof handler === "function") handler.call(this, event);
      const list = this._listeners[type];
      if (list) {
        const copy = list.slice();
        for (let i = 0; i < copy.length; i++) copy[i].call(this, event);
      }
      return true;
    }

    close() {
      this.readyState = CLOSED;
      native.close(this._handle);
    }

    _onNative(kind, payload) {
      if (this.readyState === CLOSED) return;
      if (kind === "open") {
        this.readyState = OPEN;
        this.dispatchEvent({ type: "open", target: this });
      } else if (kind === "event") {
        this.dispatchEvent({
          type: payload.type || "message",
          data: payload.data,
          lastEventId: payload.lastEventId || "",
          origin: payload.origin || "",
          target: this,
        });
      } else if (kind === "error") {
        this.readyState = (payload && payload.reconnecting) ? CONNECTING : CLOSED;
        this.dispatchEvent({ type: "error", target: this });
      }
    }
  }

  EventSource.CONNECTING = CONNECTING;
  EventSource.OPEN = OPEN;
  EventSource.CLOSED = CLOSED;
  EventSource.prototype.CONNECTING = CONNECTING;
  EventSource.prototype.OPEN = OPEN;
  EventSource.prototype.CLOSED = CLOSED;

  globalThis.EventSource = EventSource;
})();
`
